pkg/workspace_repo/api_client.go

```diff
@@ -3,74 +3,139 @@ package workspace_repo
 	"github.com/pkg/errors"
 )
 
+const (
+	tokenUnknown = 0
+	tokenPresent = 1
+	tokenMissing = 2
+)
+
 type WorkspaceRepo struct {
-	url string
-	org string
-	tfc *tfe.Client
+	url      string
+	org      string
+	hostAddr string
+	ctx      context.Context
+	tfc      *tfe.Client
 }
 
-func NewWorkspaceRepo(url string, org string) (*WorkspaceRepo, error) {
-	_, err := GetTfeToken(url, util.NewDefaultExecutor())
-	if err != nil {
-		return nil, errors.Wrap(err, "please set env var TFE_TOKEN")
-	}
-
+func NewWorkspaceRepo(ctx context.Context, url string, org string) (*WorkspaceRepo, error) {
 	// TODO do a check if see if token for the workspace repo (TFE) has expired
 	return &WorkspaceRepo{
 		url: url,
 		org: org,
+		ctx: ctx,
 	}, nil
 }
 
-func (c *WorkspaceRepo) getToken(hostname string) (string, error) {
+func (c *WorkspaceRepo) tfeLogin() error {
+	composeArgs := []string{"terraform", "login", c.hostAddr}
+
+	tf, err := exec.LookPath("terraform")
+	if err != nil {
+		return errors.Wrap(err, "terraform not in path")
+	}
+
+	cmd := &exec.Cmd{
+		Path:   tf,
+		Args:   composeArgs,
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+		Stdin:  os.Stdin,
+	}
+	err = util.NewDefaultExecutor().Run(cmd)
+	if err != nil {
+		return errors.Wrap(err, "unable to execute terraform")
+	}
+	return nil
+}
+
+func (c *WorkspaceRepo) getToken(ctx context.Context, hostname string) (string, error) {
 	// get token from env var
-	token, err := GetTfeToken(hostname, util.NewDefaultExecutor())
+	token, err := GetTfeToken(ctx, hostname)
 	if err != nil {
 		return "", errors.Wrap(err, "please set env var TFE_TOKEN")
 	}
 
 	return token, nil
 }
 
-func (c *WorkspaceRepo) getTfc() (*tfe.Client, error) {
+func (c *WorkspaceRepo) getTfc(ctx context.Context) (*tfe.Client, error) {
 	if c.tfc == nil {
 		u, err := url.Parse(c.url)
 		if err != nil {
 			return nil, err
 		}
 		hostAddr := u.Hostname()
+		c.hostAddr = hostAddr
 
-		token, err := c.getToken(hostAddr)
-		if err != nil {
-			return nil, errors.Wrap(err, "Error creating the service client:")
-		}
-		tfeConfig := &tfe.Config{
-			Address: c.url,
-			Token:   token,
-		}
-		tfc, err := tfe.NewClient(tfeConfig)
+		tfc, err := c.enforceClient(ctx)
 		if err != nil {
-			return nil, errors.Wrap(err, "Error creating the service client:")
+			c.tfc = tfc
+		} else {
+			return nil, errors.Wrap(err, "unable to create a TFE client")
 		}
-
-		c.tfc = tfc
 	}
 
 	return c.tfc, nil
 }
 
+func (c *WorkspaceRepo) enforceClient(ctx context.Context) (*tfe.Client, error) {
+	var tfc *tfe.Client
+	var err error
+	state := tokenUnknown
+
+	var token string
+	tokenPresentCounter := 0
+
+	for tokenPresentCounter < 3 {
+		switch state {
+		case tokenUnknown:
+			token, err = c.getToken(ctx, c.hostAddr)
+			if err == nil {
+				state = tokenPresent
+				tokenPresentCounter++
+			} else {
+				state = tokenMissing
+			}
+		case tokenMissing:
+			err = c.tfeLogin()
+			if err == nil {
+				state = tokenUnknown
+			}
```

Comment:
```suggestion
		case tokenMissing:
			err = c.tfeLogin()
			if err =! nil {
				errs = multierror.Append(errs, err)
				break
			}
			state = tokenUnknown
			
```
if we keep getting error in `c.tfeLogin()` will this potentiall loop forever?

Reply:
Potentially